README.md
# Oval

[![License](https://img.shields.io/badge/License-Apache_2.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![Github CI](https://github.com/qtenebrae/computer-graphics/actions/workflows/python-app.yml/badge.svg)](https://github.com/qtenebrae/computer-graphics/actions/workflows/python-app.yml)

Oval is an application designed to draw an ellipse with the possibility of moving it on the canvas. It is also possible
to fill the ellipse with different colours.

## Content

- [Intro](#intro)
- [Getting started](#getting-started)
    - [Installation](#installation)
- [Running the app](#running-the-app)
- [Project usage](#project-usage)

## Intro

The project was written as part of the Computer Graphics laboratory work.

You can see the app in action in the gif below.

<p align="center">
  <img src="https://github.com/qtenebrae/computer-graphics/blob/main/img/presentation.gif" width="600" height="500">
</p>

## Getting started

### Installation

Clone the repository

    git clone https://github.com/qtenebrae/computer-graphics.git

Switch to the repo folder

    cd computer-graphics

Installing dependencies

    pip install pytest

## Running the app

To run the application, you need to write a script:

    python .\src\__init__.py

To run the tests, it is sufficient to write the following:

    pytest

## Project usage

When the application is started for the semi-major axes A and B, the coordinates of the centre are set to default
values.

These values can be changed by the user.

The ellipse is drawn by pressing either the `Рисовать` button or the `Enter` key.

Pressing the `Up`, `Down`, `Left`, `Right` keys moves the ellipse by the specified step in the corresponding direction.

The ellipse fill colour can be changed to next / previous by pressing the `bracketleft` / `bracketleft` keys
respectively.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.1] - 2023-09-27

### Added

- Module `app`, to create and call an instance of the tkinter application.
- `app set_coords`, to set the coordinates of the ellipse on the canvas.
- `app paint`, to draw an ellipse on the canvas.
- `app on_key_press`, to process keystrokes on the keyboard.
- Module `coord`, to represent coordinates on the plane.

tests/test_app.py
import pytest

from src.app import App


class MockEvent:
    def __init__(self, keysym):
        self.keysym = keysym


@pytest.fixture(scope="module")
def app():
    app = App()
    yield app


def test_app_creation(app):
    """Тестирование создания объекта класса App."""
    assert app is not None
    assert app.title() == "Oval"
    assert app.resizable() == (False, False)


def test_set_coords(app):
    """Тестирование установки координат овала."""
    app.var_axleA.set(300)
    app.var_axleB.set(200)
    app.var_coordX.set(450)
    app.var_coordY.set(450)
    app.set_coords()
    assert app.firstCoord.x == 300
    assert app.firstCoord.y == 350
    assert app.secondCoord.x == 600
    assert app.secondCoord.y == 550


def test_paint(app):
    """Тестирование рисования овала на холсте."""
    app.var_axleA.set(300)
    app.var_axleB.set(200)
    app.var_coordX.set(450)
    app.var_coordY.set(450)
    app.currentColors = 1
    app.paint(None)
    assert app.figure is not None
    assert app.canvas.itemcget(app.figure, "fill") == "yellow"


def test_on_key_press(app):
    """Тестирование перемещения овала на холсте."""
    app.var_axleA.set(300)
    app.var_axleB.set(200)
    app.var_coordX.set(450)
    app.var_coordY.set(450)
    app.currentColors = 2
    app.paint(None)
    app.var_step.set(20)
    app.on_key_press(MockEvent("Left"))
    assert app.firstCoord.x == 280
    assert app.secondCoord.x == 580
    app.on_key_press(MockEvent("Right"))
    assert app.firstCoord.x == 300
    assert app.secondCoord.x == 600
    app.on_key_press(MockEvent("Up"))
    assert app.firstCoord.y == 330
    assert app.secondCoord.y == 530
    app.on_key_press(MockEvent("Down"))
    assert app.firstCoord.y == 350
    assert app.secondCoord.y == 550


def test_on_key_press_color(app):
    """Тестирование изменения цвета заливки овала."""
    app.var_axleA.set(300)
    app.var_axleB.set(200)
    app.var_coordX.set(450)
    app.var_coordY.set(450)
    app.currentColors = 3
    app.paint(None)
    app.on_key_press(MockEvent("bracketleft"))
    assert app.currentColors == 2
    assert app.canvas.itemcget(app.figure, "fill") == "red"
    app.on_key_press(MockEvent("bracketright"))
    assert app.currentColors == 3
    assert app.canvas.itemcget(app.figure, "fill") == "green"


main.py
from tkinter import *

import tkinter.font as tkFont


class Coord:
    def __init__(self, x, y):
        self.x = x
        self.y = y


class App(Tk):
    def __init__(self, resizeable=False):
        Tk.__init__(self)
        self.title("Oval")
        self.font = tkFont.Font(family='Bahnschrift', size=16, weight='bold')

        self.pict = Frame(self)
        self.manage = Frame(self)
        self.pict.pack(side=LEFT)
        self.manage.pack(side=RIGHT)

        self.canvas = Canvas(self.pict, width=900, height=900)
        self.canvas.create_rectangle(0, 0, 900, 900, outline='#fff', fill='#fff')
        self.canvas.pack(fill=BOTH, expand=1)

        self.var_axleA = IntVar()
        self.var_axleA.set(300)
        self.var_axleB = IntVar()
        self.var_axleB.set(200)
        self.var_coordX = IntVar()
        self.var_coordX.set(450)
        self.var_coordY = IntVar()
        self.var_coordY.set(450)
        self.var_step = IntVar()
        self.var_step.set(10)

        self.axleA = Label(self.manage, text='Полуось A', font=self.font)
        self.valueAxleA = Entry(self.manage, width=10, textvariable=self.var_axleA, font=self.font)

        self.axleB = Label(self.manage, text='Полуось B', font=self.font)
        self.valueAxleB = Entry(self.manage, width=10, textvariable=self.var_axleB, font=self.font)

        self.coordCenter = Label(self.manage, text='Координаты центра', font=self.font)
        self.coordX = Entry(self.manage, width=3, textvariable=self.var_coordX, font=self.font)
        self.coordY = Entry(self.manage, width=3, textvariable=self.var_coordY, font=self.font)

        self.step = Label(self.manage, text='Шаг', font=self.font)
        self.valueStep = Entry(self.manage, width=5, textvariable=self.var_step, font=self.font)

        self.coordCenter.pack()
        self.coordX.pack(pady=10)
        self.coordY.pack(pady=[10, 20])

        self.axleA.pack()
        self.valueAxleA.pack()
        self.axleB.pack(pady=10)
        self.valueAxleB.pack(pady=[10, 20])

        self.step.pack(pady=10)
        self.valueStep.pack(pady=[10, 20])

        self.firstCoord = Coord(0, 0)
        self.secondCoord = Coord(10, 10)
        self.colors = ["grey", "yellow", "red", "green", "blue", "pink"]
        self.currentColors = 0

        self.figure = None

        self.starterButton = Button(self.manage, text="Рисовать", font=self.font)
        self.starterButton.bind('<Button-1>', self.paint)
        self.starterButton.pack(pady=10)

        self.bind('<KeyPress>', self.on_key_press)

    def paint(self, event):
        a = int(self.valueAxleA.get())
        b = int(self.valueAxleB.get())
        x = int(self.var_coordX.get())
        y = int(self.var_coordY.get())
        self.firstCoord.x = x - a / 2
        self.firstCoord.y = y - b / 2
        self.secondCoord.x = x + a / 2
        self.secondCoord.y = y + b / 2
        if self.figure is None:
            self.figure = self.canvas.create_oval(self.firstCoord.x, self.firstCoord.y, self.secondCoord.x,
                                                  self.secondCoord.y,
                                                  fill=self.colors[self.currentColors],
                                                  outline='white')
        else:
            self.canvas.coords(self.figure, self.firstCoord.x, self.firstCoord.y, self.secondCoord.x,
                               self.secondCoord.y)

    def on_key_press(self, event):
        if self.figure is None:
            return
        step = int(self.valueStep.get())
        if event.keysym == 'Left':
            self.canvas.coords(self.figure, self.firstCoord.x - step, self.firstCoord.y, self.secondCoord.x - step,
                               self.secondCoord.y)
            self.firstCoord.x -= step
            self.secondCoord.x -= step
        elif event.keysym == 'Right':
            self.canvas.coords(self.figure, self.firstCoord.x + step, self.firstCoord.y, self.secondCoord.x + step,
                               self.secondCoord.y)
            self.firstCoord.x += step
            self.secondCoord.x += step
        elif event.keysym == 'Up':
            self.canvas.coords(self.figure, self.firstCoord.x, self.firstCoord.y - step, self.secondCoord.x,
                               self.secondCoord.y - step)
            self.firstCoord.y -= step
            self.secondCoord.y -= step
        elif event.keysym == 'Down':
            self.canvas.coords(self.figure, self.firstCoord.x, self.firstCoord.y + step, self.secondCoord.x,
                               self.secondCoord.y + step)
            self.firstCoord.y += step
            self.secondCoord.y += step
        elif event.keysym == 'bracketleft':
            if self.currentColors != 0:
                self.currentColors -= 1
            else:
                self.currentColors = len(self.colors) - 1
            self.canvas.itemconfig(self.figure, fill=self.colors[self.currentColors])
        elif event.keysym == 'bracketright':
            if self.currentColors != len(self.colors) - 1:
                self.currentColors += 1
            else:
                self.currentColors = 0
            self.canvas.itemconfig(self.figure, fill=self.colors[self.currentColors])


app = App()
app.mainloop()


src/app.py
from tkinter import *

import tkinter.font as tkFont
from coord import Coord


class App(Tk):
    """Класс для создания и вызова экземпляра приложения tkinter

    Атрибуты:
        pict (Frame): фрейм для отображения рисунка.
        manage (Frame): фрейм для управления параметрами овала.
        canvas (Canvas): холст для рисования овала.
        var_axleA (IntVar): переменная для хранения значения полуоси A.
        var_axleB (IntVar): переменная для хранения значения полуоси B.
        var_coordX (IntVar): переменная для хранения значения координаты X центра овала.
        var_coordY (IntVar): переменная для хранения значения координаты Y центра овала.
        var_step (IntVar): переменная для хранения значения шага перемещения овала.
        axleA (Label): метка для отображения текста "Полуось A".
        valueAxleA (Entry): поле для ввода значения полуоси A.
        axleB (Label): метка для отображения текста "Полуось B".
        valueAxleB (Entry): поле для ввода значения полуоси B.
        coordCenter (Label): метка для отображения текста "Координаты центра".
        coordX (Entry): поле для ввода значения координаты X центра овала.
        coordY (Entry): поле для ввода значения координаты Y центра овала.
        step (Label): метка для отображения текста "Шаг".
        valueStep (Entry): поле для ввода значения шага перемещения овала.
        firstCoord (Coord): объект класса Coord для хранения координат первой точки овала.
        secondCoord (Coord): объект класса Coord для хранения координат второй точки овала.
        colors (list): список цветов для заливки овала.
        currentColors (int): индекс текущего цвета заливки овала.
        paintButton (Button): кнопка для запуска рисования овала.
        figure (int): идентификатор фигуры на холсте.

    Методы:
        set_coords()
            Считывает параметры из переменных *Var,
            вычисляет значения x,y для точек firstCoord и secondCoord,
            сохраняет вычисленные значения.
        paint(event)
            Обрабатывает нажатие кнопки paintButton,
            рисует фигуру, либо изменяет ее местоположение.
        on_key_press(event)
            Обрабатывает нажатия клавиш на клавиатуре,
            перемещает фигуру вниз/вверх/влево/вправо на заданный шаг,
            перекрашивает фигуру на следующий/предыдущий цвет.
    """

    def __init__(self):
        Tk.__init__(self)
        self.title("Oval")
        self.resizable(False, False)
        self.font = tkFont.Font(family='Bahnschrift', size=16, weight='bold')

        self.pict = Frame(self)
        self.manage = Frame(self)
        self.pict.pack(side=LEFT)
        self.manage.pack(side=RIGHT)

        self.canvas = Canvas(self.pict, width=900, height=900)
        self.canvas.create_rectangle(0, 0, 900, 900, outline='#fff', fill='#fff')
        self.canvas.pack(fill=BOTH, expand=1)

        self.var_axleA = IntVar()
        self.var_axleA.set(300)
        self.var_axleB = IntVar()
        self.var_axleB.set(200)
        self.var_coordX = IntVar()
        self.var_coordX.set(450)
        self.var_coordY = IntVar()
        self.var_coordY.set(450)
        self.var_step = IntVar()
        self.var_step.set(10)

        self.axleA = Label(self.manage, text='Полуось A', font=self.font)
        self.valueAxleA = Entry(self.manage, width=10, textvariable=self.var_axleA, font=self.font)

        self.axleB = Label(self.manage, text='Полуось B', font=self.font)
        self.valueAxleB = Entry(self.manage, width=10, textvariable=self.var_axleB, font=self.font)

        self.coordCenter = Label(self.manage, text='Координаты центра', font=self.font)
        self.coordX = Entry(self.manage, width=3, textvariable=self.var_coordX, font=self.font)
        self.coordY = Entry(self.manage, width=3, textvariable=self.var_coordY, font=self.font)

        self.step = Label(self.manage, text='Шаг', font=self.font)
        self.valueStep = Entry(self.manage, width=5, textvariable=self.var_step, font=self.font)

        self.coordCenter.pack()
        self.coordX.pack(pady=10)
        self.coordY.pack(pady=[10, 20])

        self.axleA.pack()
        self.valueAxleA.pack()
        self.axleB.pack(pady=10)
        self.valueAxleB.pack(pady=[10, 20])

        self.step.pack(pady=10)
        self.valueStep.pack(pady=[10, 20])

        self.firstCoord = Coord(0, 0)
        self.secondCoord = Coord(10, 10)
        self.colors = ["grey", "yellow", "red", "green", "blue", "pink"]
        self.currentColors = 0

        self.paintButton = Button(self.manage, text="Рисовать", font=self.font)
        self.paintButton.bind('<Button-1>', self.paint)
        self.paintButton.pack(pady=10)

        self.bind('<KeyPress>', self.on_key_press)

        self.figure = None

    def set_coords(self):
        """Метод для установки координат овала на холсте."""

        a = int(self.valueAxleA.get())
        b = int(self.valueAxleB.get())
        x = int(self.var_coordX.get())
        y = int(self.var_coordY.get())
        self.firstCoord.x = x - a / 2
        self.firstCoord.y = y - b / 2
        self.secondCoord.x = x + a / 2
        self.secondCoord.y = y + b / 2

    def paint(self, event):
        """Метод для отрисовки овала на холсте."""

        self.set_coords()
        if self.figure is None:
            self.figure = self.canvas.create_oval(self.firstCoord.x, self.firstCoord.y, self.secondCoord.x,
                                                  self.secondCoord.y,
                                                  fill=self.colors[self.currentColors],
                                                  outline='white')
        else:
            self.canvas.coords(self.figure, self.firstCoord.x, self.firstCoord.y, self.secondCoord.x,
                               self.secondCoord.y)

    def on_key_press(self, event):
        """Метод для обработки нажатий клавиш на клавиатуре.

        Аргументы:
            event (Event): объект события, содержащий информацию о нажатой клавише.

        Действия:
            - Если фигура не создана, метод завершает работу.
            - Если нажата клавиша "Left", фигура перемещается влево на значение шага.
            - Если нажата клавиша "Right", фигура перемещается вправо на значение шага.
            - Если нажата клавиша "Up", фигура перемещается вверх на значение шага.
            - Если нажата клавиша "Down", фигура перемещается вниз на значение шага.
            - Если нажата клавиша "bracketleft", цвет заливки фигуры изменяется на предыдущий цвет из списка цветов.
            - Если нажата клавиша "bracketright", цвет заливки фигуры изменяется на следующий цвет из списка цветов.
        """

        if self.figure is None:
            if event.keysym == "Return":
                self.paint(event)
            else:
                return

        step = int(self.valueStep.get())
        if event.keysym == 'Left':
            self.canvas.coords(self.figure, self.firstCoord.x - step, self.firstCoord.y, self.secondCoord.x - step,
                               self.secondCoord.y)
            self.firstCoord.x -= step
            self.secondCoord.x -= step
        elif event.keysym == 'Right':
            self.canvas.coords(self.figure, self.firstCoord.x + step, self.firstCoord.y, self.secondCoord.x + step,
                               self.secondCoord.y)
            self.firstCoord.x += step
            self.secondCoord.x += step
        elif event.keysym == 'Up':
            self.canvas.coords(self.figure, self.firstCoord.x, self.firstCoord.y - step, self.secondCoord.x,
                               self.secondCoord.y - step)
            self.firstCoord.y -= step
            self.secondCoord.y -= step
        elif event.keysym == 'Down':
            self.canvas.coords(self.figure, self.firstCoord.x, self.firstCoord.y + step, self.secondCoord.x,
                               self.secondCoord.y + step)
            self.firstCoord.y += step
            self.secondCoord.y += step
        elif event.keysym == 'bracketleft':
            if self.currentColors != 0:
                self.currentColors -= 1
            else:
                self.currentColors = len(self.colors) - 1
            self.canvas.itemconfig(self.figure, fill=self.colors[self.currentColors])
        elif event.keysym == 'bracketright':
            if self.currentColors != len(self.colors) - 1:
                self.currentColors += 1
            else:
                self.currentColors = 0
            self.canvas.itemconfig(self.figure, fill=self.colors[self.currentColors])
